### PR TITLE
Issue103

### PR DESCRIPTION
**Pull request by EGCGDanForth**
Create an excel report containing human tissue derived samples recorded in our LIMS to meet our regulatory obligations.

The report extracts submitted samples from the lims that are still present in storage then writes the details of their location and ethics approval to an xlsx file sorted by sample name in reverse alphanumeric order. The report is then sent to the team via email.

### CODE REVIEW

File: scripts/generate_human_tissue_report.py [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+
+    # create a dictionary where key is the sample name so can sort later by reverse sample name when writing csv
+    samples_dict = {}
+    for sample in submitted_samples_batch:
+        samples_dict[sample.name] = sample
+
+    sample_keys_list = samples_dict.keys()
+
+    # sort sample keys
+    sample_keys_list = sorted(sample_keys_list, reverse=True)
```
Comment (tcezard) [inline scripts/generate_human_tissue_report.py]: You can sort the list in place with
```suggestion
    sample_keys_list.sort(reverse=True)
```

File: tests/test_generate_human_tissue_report.py [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+                              ]
+    #patch get human artifacts, patch email send and patch saving of report
+    patched_get_human_artifacts = patch('scripts.generate_human_tissue_report.get_human_artifacts', return_value=mock_submitted_samples)
+    patch_email = patch('egcg_core.notifications.email.send_email')
+    patch_workbook_save = patch('openpyxl.workbook.workbook.Workbook.save')
+
+    def setUp(self):
+        self.epp = GenerateHumanTissueReport(self.default_argv)
+
+    def test_generate_report_and_send_email(self):
```
Comment (tcezard) [inline tests/test_generate_human_tissue_report.py]: This test that the email has been sent but does not check the content of the attachment.
You should let the file be saved to disk and check the md5 (if it is deterministic).
You could also patch `Workbook` `__setitem__` (which is used by the `ws['item']`) to capture what was entered in each cell and test that.

File: scripts/generate_human_tissue_report.py (hunk 1 above)
Comment (EGCGDanForth) [reply scripts/generate_human_tissue_report.py]: I have reversed this, I get  AttributeError: 'dict_keys' object has no attribute 'sort'

File: tests/test_generate_human_tissue_report.py (hunk 2 above)
Comment (EGCGDanForth) [reply tests/test_generate_human_tissue_report.py]: I have updated to test the md5sum of the file. I have also introduced a test that tests the last instance of __setitem__. I wasn't sure how to directly test all instances of this.

Comment (mwhamgenomics) [reply tests/test_generate_human_tissue_report.py]: There are a couple of ways of doing it:

```python
# assert_any_call - assert that something has been called a certain way at some point
mocked_thing.assert_any_call('this', 'that')
mocked_thing.assert_any_call('this', 'that', other='another')

# assert_has_calls - assert that something has only had these exact calls
from unittest.mock import call
mocked_thing.assert_has_calls(
    [
        call('this', 'that'),
        call('this', 'that', other='another')
    ]
)
```

File: tests/test_generate_human_tissue_report.py [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+
+    def test_email_contents(self): #test md5sum of the excel sheet
+        fem = FakeEntitiesMaker()
+        self.epp.lims = fem.lims
+        with self.patched_get_human_artifacts, self.patch_email:
+            self.epp._run()
+
+            expected_report_md5sum = '169e0618ce9fb608a9de66a8aef8e0e0'
+            actual_report_md5sum = self.stripped_md5(self.report_path)
+            assert actual_report_md5sum == actual_report_md5sum
```
Comment (mwhamgenomics) [inline tests/test_generate_human_tissue_report.py]: This is comparing `actual_report_md5sum` with itself - we want to compare the actual and expected

File: tests/test_generate_human_tissue_report.py [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+
+
+    def test_worksheet_setitem(self): #test worksheet is populated
+        fem = FakeEntitiesMaker()
+        self.epp.lims = fem.lims
+        with self.patched_get_human_artifacts, self.patch_email, self.patch_worksheet_populate_cell as mock_populate_cell:
+            self.epp._run()
+
+            mock_populate_cell.assert_called_with('H2','SHELF1')
```
Comment (mwhamgenomics) [inline tests/test_generate_human_tissue_report.py]: We can assert more about the calls here:
```python
mock_populate_cell.assert_has_calls(
    [
        call('A1', 'PI'),
        call('B1', 'Sample Type'),
        call('C1', 'Project Name'),
        call('D1', 'Submitted Sample Name'),
        call('E1', 'REC#'),
        call('F1', 'Ethics Committee'),
        call('G1', 'Freezer'),
        call('H1', 'Shelf'),
        call('A2', 'Edinburgh Genomics'),
        call('B2', 'DNA'),
        call('C2', 'X19999'),
        call('D2', 'Sample1'),
        call('E2', 'EthicNumber1'),
        call('F2', 'Spectra'),
        call('G2', 'FREEZER1'),
        call('H2', 'SHELF1')
    ]
)
```

File: tests/test_generate_human_tissue_report.py (hunk 3 above)
Comment (EGCGDanForth) [reply tests/test_generate_human_tissue_report.py]: md5sum test removed as excel creation non-deterministic

File: tests/test_generate_human_tissue_report.py (hunk 4 above)
Comment (EGCGDanForth) [reply tests/test_generate_human_tissue_report.py]: added

**Pull request merged**